inspect/inspect.go

```diff
@@ -0,0 +1,174 @@
+package inspect
+
+import (
+	"context"
+	"errors"
+	"net"
+	"sync"
+
+	cfg "github.com/tendermint/tendermint/config"
+	inspect_rpc "github.com/tendermint/tendermint/inspect/rpc"
+	"github.com/tendermint/tendermint/libs/log"
+	tmstrings "github.com/tendermint/tendermint/libs/strings"
+	rpccore "github.com/tendermint/tendermint/rpc/core"
+	sm "github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/state/indexer"
+	"github.com/tendermint/tendermint/state/indexer/sink"
+	"github.com/tendermint/tendermint/store"
+	"github.com/tendermint/tendermint/types"
+)
+
+// Inspect manages an RPC service that exports methods to debug a failed node.
+// After a node shuts down due to a consensus failure, it will no longer start
+// up its state cannot easily be inspected. An Inspect value provides a similar interface
+// to the node, using the underlying Tendermint data stores, without bringing up
+// any other components. A caller can query the Inspect service to inspect the
+// persisted state and debug the failure.
+type Inspect struct {
```

Comment:
This isn't our convention at the moment, and may not be a great convention globally, *but* I'm a big fan of the thing, where you don't export concrete types and instead export interfaces, and keep everything else internal. I did this for a trivial type: https://github.com/tendermint/tendermint/blob/master/internal/p2p/conn_tracker.go#L10

Reply:
Unusually, I disagree: As a rule, Go packages should return _concrete_ types in almost all cases, except where for some reason they can't (e.g., a factory function that returns different implementations based on parameters). This is especially true when the interface being returned does not belong to the implementing package, where the dependency makes refactoring and static analysis more brittle.

Of course, it makes sense for the _tests_ to verify interface satisfaction, when that's what you care about, but the type signature should usually speak of the concrete type (or its pointer) rather than the interface itself.

Ref. https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Reply:
Reasonably speaking the thing that we're constructing here is roughly the same as the `node` and sort of operates like a `service.Service` which perhaps colors my impression/read of this part of the code. I don't really like the `service.Service` implementation, but I think we should keep using it, and spend a couple of days changing it all when we change it to something more reasonable later. So I guess my proposal is really "make this a `service.Service` and keep this struct private.) 

I feel like the code review/style suggestion is "you shouldn't use interfaces to mock,"  which is cool I guess, (except you have to use interfaces to use mockery, and I guess having a lot of little interfaces closer to the definition of the consuming code is... _fine_ but doesn't feel super different to me? I also guess the big benefit of organizing code "my way," is that it makes the documentation and intended use of the package/types package pretty clear and centralized (and protects against leaking implementation details when fields become public/etc. I don't feel super strongly about this, but I wish that this bit of convention was reasoned in a way that was more than just mocking?

Reply:
They call out use for mocking in that doc because that's one of the most common reasons people return interfaces rather than concrete types.  Programmatic mocks have their own cluster of problems that aren't really relevant here (though I strongly dislike them and tend to consider them a red flag for API design).

Returning a concrete type here doesn't prevent us from using the existing `Service` interface—returning concrete vs. interface is orthogonal to that. I'm not advocating we change the protocol here, I agree this isn't the time to do it.

Other advantages in my experience include:

1. Less coupling: If we return a concrete type, we don't need to import another package just to utter the name of the interface. That speeds up type checking and makes builds faster (including the build of anyone who imports this package)
2. Testing: It's nicer to write and easier to read unit tests written against the concrete implementation. Interface compliance tests still work just fine, even if the package exports its type.
3. Readability: The reader of the package that uses the implementation sees the actual type they're instantiating, rather than having to chase it down through constructors or factory functions.
4. Usability: Often the user of the package cares about the _specific_ implementation they're using (e.g., "I want the memory-backed KV implementation so I can build a cache"), and exporting the concrete type allows you to expose other functionality besides just what's in the interface. Even if you don't care about that at the moment you implement it, it's hard to go back and fix it later without adding type assertions and other fragile workarounds.

All of which is to say: The advice to return concrete things is a good default. It's not a hard-and-fast rule and there are (as you mentioned) some reasons to do otherwise. I'd argue, however, that we shouldn't do it just to "hide" the name of a type from the user of the package. The purpose of this package is to provide a _special_ implementation of the interface that has _different_ constraints, and I don't see much practical benefit in cosmetically obscuring that from the user of the package.

Reply:
I'm not completely certain I understand the concern about leaking implementation details:
> and protects against leaking implementation details when fields become public/etc

If we make the fields public, it's because we think they should be publically viewable/modifiable. I'm not sure having the constructor be the place we determine what should be visible is ideal.

In terms of mocking, I find the pattern of letting the consumer write interfaces to work reasonably well. It allows the consumer to provide a much more clear contract against the packages it uses. The consumer requires a type that can `DoThing()` and any type that can `DoThing()` works. If we can easily make it look like a `service.Service` I think it's fine to do that where it is used. That way, if the caller wants to use the type with something that takes a `service.Serivce`, the caller is responsible for shimming the types together. The caller has to manage its dependency with the `serivce` package already in that case so managing how the `inspect` and `service` interact becomes a natural extension of managing the dependency. Additionally, changes to the `service.Service` interface do not become the problem of the `inspect.Inspect` package. Anyone using the `inspect` package can continue to use it as normal if the `serivce.Service` changes.

Ultimately, we want an `inspect.Inspect` to be invoked in the `cmd` package when we go to run when of these, not any old `service.Service` so we definitely care about which type we're using. Using the concrete type means that the contract is more clear imo. The `inspect` package provides a type that looks one way and has methods `X()` `Y()` and `Z()`. The package can't break it's contract with the code that consumes it as easily.